File: backend/src/modules/htable/htable.controller.ts
import { Controller, Get, Post, Body, Patch, Param, Delete, Query, ParseIntPipe } from '@nestjs/common';
import { ApiTags, ApiOperation } from '@nestjs/swagger';
import { HtableService } from './htable.service';
import { PaginationDto } from '../../common/dto/pagination.dto';
import { ApiResponseDto } from '../../common/dto/api-response.dto';

@ApiTags('哈希表')
@Controller('api/htable')
export class HtableController {
  constructor(private readonly htableService: HtableService) {}

  @Get()
  @ApiOperation({ summary: '获取哈希表记录（数据库）' })
  async findAll(@Query() dto: PaginationDto) {
    return ApiResponseDto.success(await this.htableService.findAll(dto));
  }

  @Post()
  async create(@Body() data: any) {
    return ApiResponseDto.success(await this.htableService.create(data), '创建成功');
  }

  @Patch(':id')
  async update(@Param('id', ParseIntPipe) id: number, @Body() data: any) {
    return ApiResponseDto.success(await this.htableService.update(id, data), '更新成功');
  }

  @Delete(':id')
  async remove(@Param('id', ParseIntPipe) id: number) {
    await this.htableService.remove(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  // RPC 操作
  @Get('rpc/get')
  @ApiOperation({ summary: '获取内存值' })
  async getValue(@Query('table') table: string, @Query('key') key: string) {
    return ApiResponseDto.success(await this.htableService.getValue(table, key));
  }

  @Post('rpc/sets')
  @ApiOperation({ summary: '设置字符串值' })
  async setString(@Body() body: { table: string; key: string; value: string }) {
    await this.htableService.setString(body.table, body.key, body.value);
    return ApiResponseDto.success(null, '设置成功');
  }

  @Post('rpc/seti')
  @ApiOperation({ summary: '设置整数值' })
  async setInt(@Body() body: { table: string; key: string; value: number }) {
    await this.htableService.setInt(body.table, body.key, body.value);
    return ApiResponseDto.success(null, '设置成功');
  }

  @Post('rpc/delete')
  @ApiOperation({ summary: '删除键' })
  async deleteKey(@Body() body: { table: string; key: string }) {
    await this.htableService.deleteKey(body.table, body.key);
    return ApiResponseDto.success(null, '删除成功');
  }

  @Get('rpc/dump')
  @ApiOperation({ summary: '导出表' })
  async dump(@Query('table') table: string) {
    return ApiResponseDto.success(await this.htableService.dump(table));
  }

  @Post('rpc/reload')
  @ApiOperation({ summary: '重载表' })
  async reload(@Body() body: { table: string }) {
    await this.htableService.reload(body.table);
    return ApiResponseDto.success(null, '重载成功');
  }

  @Get('rpc/tables')
  @ApiOperation({ summary: '列出所有表' })
  async listTables() {
    return ApiResponseDto.success(await this.htableService.listTables());
  }
}

File: kam-admin-module-session/src/main/java/io/kamailio/admin/modules/htable/controller/HtableController.java
package io.kamailio.admin.modules.htable.controller;

import io.kamailio.admin.common.dto.ApiResponse;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.htable.entity.Htable;
import io.kamailio.admin.modules.htable.service.HtableService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Tag(name = "哈希表")
@RestController
@RequestMapping("/api/htable")
public class HtableController {
    private final HtableService service;

    public HtableController(HtableService service) { this.service = service; }

    @GetMapping
    @Operation(summary = "获取哈希表记录（数据库）")
    public ResponseEntity<ApiResponse<PaginatedResult<Htable>>> findAll(PaginationDto dto) {
        return ResponseEntity.ok(ApiResponse.success(service.findAll(dto)));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Htable>> create(@RequestBody Htable data) {
        return ResponseEntity.ok(ApiResponse.success(service.create(data), "创建成功"));
    }

    @PatchMapping("/{id}")
    public ResponseEntity<ApiResponse<Htable>> update(@PathVariable Integer id, @RequestBody Htable data) {
        return ResponseEntity.ok(ApiResponse.success(service.update(id, data), "更新成功"));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> remove(@PathVariable Integer id) {
        service.remove(id);
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @GetMapping("/rpc/get")
    @Operation(summary = "获取内存值")
    public ResponseEntity<ApiResponse<Object>> getValue(@RequestParam String table, @RequestParam String key) {
        return ResponseEntity.ok(ApiResponse.success(service.getValue(table, key)));
    }

    @PostMapping("/rpc/sets")
    @Operation(summary = "设置字符串值")
    public ResponseEntity<ApiResponse<Void>> setString(@RequestBody Map<String, String> body) {
        service.setString(body.get("table"), body.get("key"), body.get("value"));
        return ResponseEntity.ok(ApiResponse.success(null, "设置成功"));
    }

    @PostMapping("/rpc/seti")
    @Operation(summary = "设置整数值")
    public ResponseEntity<ApiResponse<Void>> setInt(@RequestBody Map<String, Object> body) {
        service.setInt((String) body.get("table"), (String) body.get("key"), ((Number) body.get("value")).intValue());
        return ResponseEntity.ok(ApiResponse.success(null, "设置成功"));
    }

    @PostMapping("/rpc/delete")
    @Operation(summary = "删除键")
    public ResponseEntity<ApiResponse<Void>> deleteKey(@RequestBody Map<String, String> body) {
        service.deleteKey(body.get("table"), body.get("key"));
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @GetMapping("/rpc/dump")
    @Operation(summary = "导出表")
    public ResponseEntity<ApiResponse<Object>> dump(@RequestParam String table) {
        return ResponseEntity.ok(ApiResponse.success(service.dump(table)));
    }

    @PostMapping("/rpc/reload")
    @Operation(summary = "重载表")
    public ResponseEntity<ApiResponse<Void>> reload(@RequestBody Map<String, String> body) {
        service.reload(body.get("table"));
        return ResponseEntity.ok(ApiResponse.success(null, "重载成功"));
    }

    @GetMapping("/rpc/tables")
    @Operation(summary = "列出所有表")
    public ResponseEntity<ApiResponse<Object>> listTables() {
        return ResponseEntity.ok(ApiResponse.success(service.listTables()));
    }

    @GetMapping("/rpc/stats")
    @Operation(summary = "获取哈希表统计")
    public ResponseEntity<ApiResponse<Object>> stats() {
        return ResponseEntity.ok(ApiResponse.success(service.stats()));
    }
}
